### PR TITLE
Fix generic constraint handling in IDE0041

### LINKED ISSUE #47972: "IDE0041 Null check can be simplified" ignores generic type constraint when deciding between 'is null' and '== null'.
**Steps to Reproduce**:

1. Define a class with an overridden `==` operator:
```csharp
public class Example
{
    public static bool operator ==(Example left, Example right) => true;
    public static bool operator !=(Example left, Example right) => !(left == right);
    public override bool Equals(object obj) => true;
    public override int GetHashCode() => 0;
}
```

2. Define a null-checking method, with a generic parameter constrained to the `Example` class:
```csharp
public static bool IsNull<T>(T arg) where T : Example
{
    return ReferenceEquals(arg, null);
}
```

3. Apply the IDE0041 suggested fix to it.

**Expected Behaviour**:

The fix should use `is null`, as it does if the method is not generic and the argument is typed as `Example`:
```csharp
public static bool IsNull(Example arg)
{
    return arg is null;
}
```
'Compiling' the C# using SharpLab shows that this is interpreted as `(object)arg == null`, which avoids using the overridden `==` operator.

**Actual Behaviour**

```csharp
public static bool IsNull<T>(T arg) where T : Example
{
    return arg == null;
}
```
The fix uses `== null`, which is a call to the overridden `==` operator and not equivalent (especially in this slightly contrived case) to `ReferenceEquals(arg, null)`.

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Fixes #47972

### CODE REVIEW

File: src/Analyzers/CSharp/Tests/UseIsNullCheck/UseIsNullCheckForReferenceEqualsTests.cs [hunk 1]
```diff
@@ -500,5 +501,67 @@ void M(string s2)
...
+        [WorkItem(23581, "https://github.com/dotnet/roslyn/issues/47972")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        public async Task TestValueParameterTypeIsBaseConstraintGeneric()
+        {
+            await TestInRegularAndScript1Async(
+@"using System;
+
+class C
+{
+    public static void NotNull<T>(T value) where T:C
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseIsNullCheck/UseIsNullCheckForReferenceEqualsTests.cs:514]: this doesn't appear to be a value-typeparameter.  i.e. it doesn't have the 'struct' constraint.

**Review by CyrusNajmabadi — APPROVED** [review]
code change lgtm, we made need more tests/test-clarity here.

File: src/Analyzers/CSharp/Tests/UseIsNullCheck/UseIsNullCheckForReferenceEqualsTests.cs [hunk 2]
```diff
@@ -500,5 +501,67 @@ void M(string s2)
...
+{
+    public static void NotNull<T>(T value)
+    {
+        if (value is null)
+        {
+            return;
+        }
+    }
+}", new TestParameters(parseOptions: CSharp8));
+        }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseIsNullCheck/UseIsNullCheckForReferenceEqualsTests.cs:565]: can you add an unconstrained test for C#7 as well?

File: src/Analyzers/CSharp/Tests/UseIsNullCheck/UseIsNullCheckForReferenceEqualsTests.cs (hunk 1 above)
Comment (huoyaoyuan) [reply src/Analyzers/CSharp/Tests/UseIsNullCheck/UseIsNullCheckForReferenceEqualsTests.cs:514]: The name is unclear. The purpose is "no `class` constraint, but has base type (class) constraint"

File: src/Analyzers/CSharp/Tests/UseIsNullCheck/UseIsNullCheckForReferenceEqualsTests.cs (hunk 2 above)
Comment (huoyaoyuan) [reply src/Analyzers/CSharp/Tests/UseIsNullCheck/UseIsNullCheckForReferenceEqualsTests.cs:565]: That's the one I added `parseOptions: CSharp7` above. I should move them together.

**Comment by huoyaoyuan** [comment]
The new test names should be consistent with existing ones now.

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**